### ISSUE TITLE
Viewer toggle mode

### ISSUE THREAD

**Issue by ieivanov** [napari/napari#2166]
## 🚀 Feature
It would be useful to be able to toggle between different image layers by clicking on the image layer.

## Motivation

The viewer currently has a `stack` mode where images in different layers can be blended together, and a `grid` mode where images in different layers are displayed side-by-side.

## Pitch

It would be useful to be able to compare features across different image layers in the same position of the ND dataset by toggling between the layers, displayed at full opacity, by clicking on the desired layer. In this new `toggle` viewer mode, only the selected layer is visible, and all other layers are hidden.

## Alternatives

Currently, this can be achieved by first hiding all layers using the eye icon. Then one layer is made visible. To make another layer visible, the user will first hide the old layer, and then unhide the new layer. This process is not convenient as it involves a lot of mouse clicks.

**Comment by sofroniewn** [comment]
> It would be useful to be able to toggle between different image layers by clicking on the image layer.

Maybe this could be enabled by clicking with a modifier key. Shift/ ctrl click will show a layer and hide all others? Curious how @jni or @royerloic would feel about this?

**Comment by jni** [comment]
I'd be ok with that shortcut!

**Comment by carsen-stringer** [comment]
I like that idea @sofroniewn, it could also be a modifier on the arrow key given you can already use the arrow key to go through the layers? But I guess it would have to be something like ALT to avoid conflicting with the selection feature

**Comment by sofroniewn** [comment]
> I like that idea @sofroniewn, it could also be a modifier on the arrow key given you can already use the arrow key to go through the layers? But I guess it would have to be something like ALT to avoid conflicting with the selection feature

Sort of like an "exclusive select"? I'd be open to that too - i might need to play with it to see exactly what it was like, but I like the idea

**Comment by ieivanov** [comment]
Thanks folks, it would be great to see something along these lines!

**Comment by haesleinhuepf** [comment]
Hi all,

I just had a similar request from my students. It would be cool to have a right-click menu on layers (because not everyone likes using short-cuts) with actions such as:
* Make all layers above invisible
* Make all layers invisible but this

**Comment by sofroniewn** [comment]
I believe @tlambert03 has begun work on this here #2556!!

**Comment by jnahlers** [comment]
I'm also really interested in this. I often have to judge whether I can see subtle differences between two different images, and so need to be able to quickly toggle back and forth between the two. Currently the only way is to open the two images and then combine them into a stack. It would be great to be able to toggle between layers as well!

Adding an `alt+` modifier to layer selection changes (whether by clicking or arrow up/down) that first changes the selected layer (or layers, if holding `shift`) and then calls some kind of `show only selected` function is probably the easiest option (mirroring [photoshop's behavior](https://graphicdesign.stackexchange.com/a/13642)).

**Comment by Czaki** [comment]
I'm not sure if it has been released, but on the current main, there is a shortcut, `V`, on the toggle visibility of selected layers.

So you could select these two layers and toggle its visibility by pressing V. If one is visible and the second is hidden, then it will behave like you describe.

**Comment by psobolewskiPhD** [comment]
@jnahlers I've implemented two feature that help address this:
Alt-click on the visibility eye to show just that one layer (hides others)
https://github.com/napari/napari/pull/5618
Expanded visibility contextual menu (right click on layer) with ability to show/hide selected or unselected layers:
https://github.com/napari/napari/pull/5574
How do those sound? Does it meet your need? can this issue be considered addressed?
Only thing is they're not released. They're slated for 0.5.0 as feature. We've been focused on bugs in 0.4.18 and 0.4.19.

**Comment by jnahlers** [comment]
@Czaki That's a neat trick! Though of course it only works for two states. If you want to rapidly flick through many layers, with only one visible at a time, this doesn't help.

@psobolewskiPhD I missed those PRs, look good. Note that I didn't open this issue - @ieivanov, would you say those PRs fulfill your needs?

For my case, I would still love something like an `alt+` modifier when changing selected layer to only have visible the new selection, as I described above. The reason is that I need to quickly flip through two or more images, and that number of clicks isn't fast enough. 

As I describe in [this issue](https://github.com/napari/napari/issues/6514), my usual solution is to make the images into a stack and scroll back and forth. But that loses the metadata of the layers, and it feels like an unnecessary step to do something which should be possible with just layers, by up and down arrowing in the layer list, together with a hypothetical 'only selected visible' modifier on.

**Comment by jnahlers** [comment]
I made myself a hacky little widget plugin that implements the kind of 'only selected visible' mode that we have been discussing:
```
def _only_selected_visible_widget_init(widget):
    @widget.show_only_selected.changed.connect
    @widget.viewer.value.layers.selection.events.changed.connect
    def update_visibility_on_selection_change(event):
        if widget.show_only_selected.value == True:
            selected_layers = widget.viewer.value.layers.selection
            for layer in widget.viewer.value.layers:
                if layer in selected_layers:
                    layer.visible = True
                else:
                    layer.visible = False
    return

@magic_factory(
    widget_init=_only_selected_visible_widget_init,
    auto_call=True,
    show_only_selected={"widget_type": "CheckBox", "value": False}
)
def only_selected_visible_widget(
        viewer: "napari.viewer.Viewer",
        show_only_selected: "bool",
) -> "bool":
    return show_only_selected
```

**Comment by ieivanov** [comment]
@Czaki thank you for reviving this issue. I personally got comfortable with the `V` shortcut, but I'm excited to see the new features that @psobolewskiPhD introduced, they offer a lot more flexibility when toggling through datasets. I consider this issue resolved, you're welcome to close it with #6590. Thanks!
